In [1]:
import pandas as pd

In [2]:
df_fusionné = pd.read_csv('df_fusionné.csv')

In [3]:
df_fusionné_binaire = df_fusionné.sample(frac=0.2)
df_fusionné_binaire.iloc[:, 0] = df_fusionné_binaire.iloc[:, 0].apply(lambda x: 0 if x == 'N' else 1)

In [4]:
# Count the occurrences of each class
counts = df_fusionné_binaire.iloc[:, 0].value_counts()

# Find the number of instances in the minority class
min_count = counts.min()

# Create balanced DataFrame through undersampling
df_balanced_under = pd.concat([
    df_fusionné_binaire[df_fusionné_binaire.iloc[:, 0] == 0].sample(min_count),
    df_fusionné_binaire[df_fusionné_binaire.iloc[:, 0] == 1].sample(min_count)
])

# Shuffle the DataFrame to mix the classes
df_balanced_under = df_balanced_under.sample(frac=1)


In [5]:
# Count occurrences of 0 and 1 in the first column
value_counts = df_balanced_under.iloc[:, 0].value_counts()

# Display the counts
print(value_counts)

ColumnName
1    7604
0    7604
Name: count, dtype: int64


In [6]:
df_balanced_under

ColumnName         0         1         2         3         4         5  \
33954           1  0.548607  0.550231  0.563333  0.556666  0.565525  0.575011   
30865           1  0.560332  0.560726  0.566756  0.563349  0.566255  0.569216   
45214           0  0.475101  0.475128  0.475157  0.475184  0.475210  0.475235   
30986           1  0.509513  0.508609  0.520137  0.504828  0.512008  0.520775   
40320           0  0.458797  0.458048  0.457523  0.457259  0.456905  0.456571   
...           ...       ...       ...       ...       ...       ...       ...   
15062           0  0.505344  0.505931  0.511825  0.510673  0.513251  0.515332   
131354          0  0.459731  0.417111  0.411450  0.410388  0.415455  0.423376   
57165           0  0.500663  0.500270  0.499941  0.499684  0.499432  0.499203   
206259          1  0.503939  0.503476  0.502943  0.502357  0.501821  0.501276   
139418          0  0.480053  0.483514  0.487190  0.492158  0.501702  0.513429   

               6         7         8  ...       422       423       424  \
33954   0.576520  0.575748  0.580219  ...  0.441659  0.446203  0.443582   
30865   0.568193  0.566213  0.568187  ...  0.514554  0.516894  0.514996   
45214   0.475259  0.475282  0.475311  ...  0.474831  0.474855  0.474876   
30986   0.518593  0.513735  0.517236  ...  0.371087  0.377619  0.373055   
40320   0.456437  0.456338  0.455959  ...  0.479186  0.480092  0.478975   
...          ...       ...       ...  ...       ...       ...       ...   
15062   0.514443  0.512511  0.512529  ...  0.469807  0.471564  0.470293   
131354  0.437908  0.453644  0.457051  ...  0.454816  0.455017  0.453656   
57165   0.499033  0.498886  0.498689  ...  0.508867  0.508111  0.507321   
206259  0.500711  0.500122  0.499432  ...  0.505118  0.504954  0.504872   
139418  0.524986  0.536903  0.550052  ...  0.490701  0.492871  0.491664   

             425       426       427       428       429       430       431  
33954   0.438637  0.441146  0.446934  0.441482  0.441334  0.447195  0.440631  
30865   0.512284  0.509588  0.507408  0.511318  0.520650  0.519867  0.509561  
45214   0.474898  0.474920  0.474945  0.474974  0.475005  0.475038  0.475071  
30986   0.365110  0.365569  0.370265  0.364627  0.367028  0.376962  0.369986  
40320   0.476922  0.474741  0.472234  0.469407  0.466555  0.463325  0.460249  
...          ...       ...       ...       ...       ...       ...       ...  
15062   0.468447  0.466785  0.465519  0.468385  0.475269  0.474347  0.466220  
131354  0.451964  0.447767  0.443685  0.444314  0.446130  0.453235  0.455349  
57165   0.506496  0.505701  0.504897  0.503982  0.503067  0.502115  0.501217  
206259  0.504820  0.504734  0.504642  0.504577  0.504489  0.504415  0.504292  
139418  0.489153  0.486671  0.484090  0.481813  0.480068  0.478773  0.478346  

[15208 rows x 433 columns]

In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split

X = df_balanced_under.iloc[:, 1:]  
y = df_balanced_under.iloc[:, 0] 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [8]:
print(y_train.unique())

[0 1]


In [9]:
y_train = y_train.astype(int)
y_test = y_test.astype(int)

In [10]:
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Conv1D, BatchNormalization, MaxPooling1D, Dropout, Flatten, Dense
from keras.optimizers import Adam

Model = Sequential()

Model.add(Conv1D(256, 10, strides=1, padding="same", activation="relu", input_shape=(432, 1)))
Model.add(BatchNormalization())
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(256, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(128, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(64, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(32, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Flatten())
Model.add(Dense(1024, activation='relu'))
Model.add(Dropout(0.3))

Model.add(Dense(units=1, activation='sigmoid'))  # Use 'sigmoid' for binary classification


/Users/arthurdesmazures/venv-metal/lib/python3.10/site-packages/keras/src/layers/convolutional/base_conv.py:99: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(
2024-03-13 00:31:05.123125: I metal_plugin/src/device/metal_device.cc:1154] Metal device set to: Apple M3 Pro
2024-03-13 00:31:05.123141: I metal_plugin/src/device/metal_device.cc:296] systemMemory: 18.00 GB
2024-03-13 00:31:05.123144: I metal_plugin/src/device/metal_device.cc:313] maxCacheSize: 6.00 GB
2024-03-13 00:31:05.123159: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:305] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2024-03-13 00:31:05.123169: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:271] Created TensorFlow device (/job:lo

In [11]:
print(Model.summary())

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d (Conv1D)                 │ (None, 432, 256)       │         2,816 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization             │ (None, 432, 256)       │         1,024 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d (MaxPooling1D)    │ (None, 216, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_1 (Conv1D)               │ (None, 216, 256)       │       655,616 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 216, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_1 (MaxPooling1D)  │ (None, 108, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_2 (Conv1D)               │ (None, 108, 128)       │       327,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 108, 128)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_2 (MaxPooling1D)  │ (None, 54, 128)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_3 (Conv1D)               │ (None, 54, 64)         │        81,984 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 54, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_3 (MaxPooling1D)  │ (None, 27, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_4 (Conv1D)               │ (None, 27, 32)         │        20,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 27, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_4 (MaxPooling1D)  │ (None, 14, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 448)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 1024)           │       459,776 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_4 (Dropout)             │ (None, 1024)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │         1,025 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,550,561 (5.91 MB)

 Trainable params: 1,550,049 (5.91 MB)

 Non-trainable params: 512 (2.00 KB)

None


In [12]:
# Define callbacks
Early_Stopper = tf.keras.callbacks.EarlyStopping(monitor="loss", patience=3, mode="min")


In [13]:
# Compile the model
Model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])

In [14]:
Conv1D_Model = Model.fit(X_train, y_train, epochs=70, validation_data=(X_test, y_test), callbacks=[Early_Stopper])

Epoch 1/70


2024-03-13 00:31:05.809814: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] Plugin optimizer for device_type GPU is enabled.


352/381 ━━━━━━━━━━━━━━━━━━━━ 0s 34ms/step - accuracy: 0.7733 - loss: 0.4576

KeyboardInterrupt: 

In [ ]:
import pickle

with open('Conv1D_Model.pkl', 'wb') as file:
    pickle.dump(Conv1D_Model, file)

print("Model saved successfully.")

AttributeError: Can't pickle local object 'Layer._initializer_tracker.<locals>.<lambda>'

In [15]:
Model_Results = Model.evaluate(X_test,y_test)
print("LOSS:  " + "%.4f" % Model_Results[0])
print("ACCURACY:  " + "%.4f" % Model_Results[1])

95/95 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.8320 - loss: 0.3854
LOSS:  0.3815
ACCURACY:  0.8255


In [16]:
# Assuming your model is named 'Model'
y_pred_probs = Model.predict(X_test)
y_pred = (y_pred_probs > 0.5).astype("int32")  # Convert probabilities to binary labels


95/95 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step


In [17]:
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score, roc_curve, log_loss

# Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)

# Precision
precision = precision_score(y_test, y_pred)
print(f"Precision: {precision:.2f}")

# Recall
recall = recall_score(y_test, y_pred)
print(f"Recall: {recall:.2f}")

# F1 Score
f1 = f1_score(y_test, y_pred)
print(f"F1 Score: {f1:.2f}")

# ROC-AUC Score
roc_auc = roc_auc_score(y_test, y_pred_probs)
print(f"ROC AUC Score: {roc_auc:.2f}")

# Now, you can calculate the log loss
logloss = log_loss(y_test, y_pred_probs)
print(f"Log Loss: {logloss:.2f}")

Confusion Matrix:
 [[1530   13]
 [ 514  963]]
Precision: 0.99
Recall: 0.65
F1 Score: 0.79
ROC AUC Score: 0.95
Log Loss: 0.38


Confusion Matrix:
 [[1456  106]
 [ 640  834]]
Precision: 0.89
Recall: 0.57
F1 Score: 0.69
ROC AUC Score: 0.86
Log Loss: 0.65

Confusion Matrix:
 [[1357  127]
 [ 325 1187]]
Precision: 0.90
Recall: 0.79
F1 Score: 0.84
ROC AUC Score: 0.95
Log Loss: 0.33

Confusion Matrix:
 [[1401  104]
 [ 947  576]]
Precision: 0.85
Recall: 0.38
F1 Score: 0.52
ROC AUC Score: 0.81
Log Loss: 0.64
